### ISSUE TITLE
Steam reverts even when it doesn't

### ISSUE THREAD

**Issue by Foggalong** [Foggalong/hardcode-fixer#57]
The program output states it's reverting Steam icons, even when it doesn't actually do anything

**Comment by hlechner** [comment]
An possible way is to apply [chattr](http://linux.die.net/man/1/chattr) attribute on .desktop file as root so the application can't revert/update the file.

But of course, blocking the steam to update the .desktop files is not a real good solution.

**Comment by hlechner** [comment]
@Foggalong I'm not a programmer but I've made some changes into script to add the ability to "lock" the .desktop by using "-k" or "--lock".

Take a look on my fork:
https://github.com/hlechner/hardcode-fixer/commit/15761804bd2123e4279d01232a0c0850675e5471

**Comment by Foggalong** [comment]
Huh, that's an interesting way of fixing the updating issue. I might look into merging that in when I have a bit more time :)

That isn't the problem with this issue though, as far as I can tell it's simply a ghost message: it says it's reverting when it doesn't need to, but it doesn't actually change anything anyway.

**Comment by hlechner** [comment]
Now i got it. Yea, I misunderstood it. I'll try to replicate the same issue later today and if I find the solution I send you a Pull Request

**Comment by Foggalong** [comment]
Thanks!

**Comment by hlechner** [comment]
I couldn't replicate the very same behavior as you are facing. I'm afraid of that it can be some environment related issue.

Here my step-by-step to reproduce the issue:
**1** - Open **Steam** client, go to **Dota 2** Game, Right click > "Create Desktop Shortcut" 
**2** - Copy **Dota 2.desktop** from ``~/Desktop/`` to ``~/.local/share/applications/``
**3** - Execute the script to fix it as normal user (non-root):
```
$ hardcode-fixer
The script must be run as root to (un)fix global launchers.
Do you want to continue in local mode? yes
S: Fixing Dota 2...
``` 
**4** - Check if the desktop file was changed: Yes!
```
$ cat ~/.local/share/applications/Dota\ 2.desktop | grep Icon
Icon=steam_icon_570
```
**5** - Try to execute again the script with the desktop file already changed:
```
$ hardcode-fixer
The script must be run as root to (un)fix global launchers.
Do you want to continue in local mode? yes
S: Fixing Dota 2...
```
```
$ cat ~/.local/share/applications/Dota\ 2.desktop | grep Icon
Icon=steam_icon_570
```
**6** - Run the script with reverse mode ( -r ):
```
$ hardcode-fixer -r
This will undo all changes previously made.
Are you sure you want to continue? yes
The script must be run as root to (un)fix global launchers.
Do you want to continue in local mode? yes
S: Reverting Dota 2...
```
```
$ cat ~/.local/share/applications/Dota\ 2.desktop | grep Icon
Icon=steam
```
**7** - Try to run it again:
```
$ hardcode-fixer -r
This will undo all changes previously made.
Are you sure you want to continue? yes
The script must be run as root to (un)fix global launchers.
Do you want to continue in local mode? yes
```

So I couldn't find the same behavior as you, could you send me your step-by-step to reproduce it? So I can go further.

Here it's my Environment:
**Distro**: Arch Linux 64-bits
**DE**: Gnome 3.14.2

**Comment by hlechner** [comment]
On my last comment (step 5) it's possible to see that the script tries to fix the desktop file even if its already fixed.

So I've investigated and the problem is inside the line 153 inside fix.sh:
```
if grep -Fq "Icon=$current" "$local_apps$launcher"; then
```

Because in both situations it will return true, it's because both start with the same string ``Icon=steam``.

A solution for this is to use **regular expression** on grep instead of **fixed string** and add an end of line.

I just need to test how to scape the variables, because in regular expression a "$" means end of line.

**Comment by Foggalong** [comment]
This issue was actually fixed in 1769a99d5c9de025a53df5ddaf296b245fe784f2, so thanks! :dancer:
